helios-services/src/main/java/com/spotify/helios/master/ZooKeeperMasterModel.java

```diff
@@ -448,36 +551,58 @@ public void updateDeploymentGroupHosts(final String groupName, final List<String
         version = Optional.of(node.getStat().getVersion());
         curHosts = Json.read(node.getBytes(), new TypeReference<List<String>>() {});
       } catch (JsonMappingException e) {
-        curHosts = Collections.emptyList();
+        curHosts = emptyList();
       } catch (NoNodeException e) {
         // DG was deleted -- abort.
         return;
       }
 
       if (!version.isPresent() || !hosts.equals(curHosts)) {
         // Node not present or hosts have changed
-        final List<ZooKeeperOperation> ops = Lists.newArrayList();
-        ops.add(set(Paths.statusDeploymentGroupHosts(groupName), Json.asBytes(hosts)));
+        final List<ZooKeeperOperation> operations = Lists.newArrayList();
+        operations.add(set(Paths.statusDeploymentGroupHosts(groupName), Json.asBytes(hosts)));
 
-        final DeploymentGroup deploymentGroup = getDeploymentGroup(groupName);
         ImmutableList<Map<String, Object>> events = ImmutableList.of();
 
-        if (deploymentGroup.getJobId() != null) {
-          final DeploymentGroupStatus deploymentGroupStatus = getDeploymentGroupStatus(groupName);
-          if (deploymentGroupStatus == null || deploymentGroupStatus.getState() != FAILED) {
-            final RollingUpdateOp op =
-                getInitRollingUpdateOps(deploymentGroup, hosts, HOSTS_CHANGED, client);
-            ops.addAll(op.operations());
-            events = op.events();
-          }
+        final RollingOperation lastGoodOp = getLastGoodRollingOperation(client, groupName);
```

Comment:
[Under the assumption that the last good rolling operation == the last op that succeeded]

IMO, hosts changing should cause the current target version to be deployed, not the last succeesfully deployed version.

For example, if I deploy version B but the rollout fails on some random host. Then when I add hosts I want version B to be deployed on them, not some older version A.